File: 2021/Day5/Word Search/Word_search.py
from collections import Counter, defaultdict


class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        char_dict, rows, cols = defaultdict(list), len(board), len(board[0])

        # keep a dict for all position of a letter
        for i in range(rows):
            for j in range(cols):
                char_dict[board[i][j]].append([i, j])

        # check if dfs should be performed
        cnt_chars = Counter(word)
        for x, cnt in cnt_chars.items():
            if cnt > len(char_dict[x]):
                return False

        def dfs(board, word, prev_x, prev_y, idx, visited):
            letter = word[idx]
            if letter in char_dict:
                # fetch list of all position
                list_pos = char_dict[letter]
                for pos in list_pos:
                    xi, yi = pos[0], pos[1]
                    if (xi, yi) not in visited and (abs(xi - prev_x) + abs(yi - prev_y)) <= 1:
                        if idx + 1 == len(word):
                            return 1
                        else:
                            ans = dfs(board, word, xi, yi, idx +
                                      1, visited + [(xi, yi)])
                            if ans:
                                return 1

            return 0

        if len(word) == 1 and word[0] in char_dict:
            return True
        if word[0] in char_dict:
            for pos in char_dict[word[0]]:
                dfs_ans = dfs(board, word, pos[0], pos[1], 1, [
                              (pos[0], pos[1])])
                if dfs_ans == 1:
                    return True
        return False

File: 2021/Day5/Word Search/Word_search.cpp
class Solution
{
public:
    bool checkPresent(vector<vector<char>> board, string word)
    {
        int A[26];
        memset(A, 0, sizeof(A));
        for (int i = 0; i < board.size(); ++i)
        {
            for (int j = 0; j < board[0].size(); ++j)
            {
                board[i][j] = toupper(board[i][j]);
                A[board[i][j] - 'A']++;
            }
        }

        for (int i = 0; i < word.length(); ++i)
        {
            word[i] = toupper(word[i]);
            A[word[i] - 'A']--;
            if (A[word[i] - 'A'] < 0)
                return false;
        }
        return true;
    }

    bool isNextValid(vector<vector<char>> board, int x, int y, string word, int count)
    {

        if (board[x][y] == word[count])
        {
            return true;
        }
        return false;
    }

    void findPath(vector<vector<char>> board, string word, bool &flag, vector<vector<int>> &visited, int count, int x, int y)
    {

        if (count == word.length() - 1)
        {
            flag = 1;
            return;
        }

        visited[x][y] = 1;

        count++;

        if (x + 1 < board.size() && !visited[x + 1][y] && isNextValid(board, x + 1, y, word, count))
        {
            findPath(board, word, flag, visited, count, x + 1, y);
        }
        if (x - 1 >= 0 && !visited[x - 1][y] && isNextValid(board, x - 1, y, word, count))
        {

            findPath(board, word, flag, visited, count, x - 1, y);
        }
        if (y + 1 < board[0].size() && !visited[x][y + 1] && isNextValid(board, x, y + 1, word, count))
        {

            findPath(board, word, flag, visited, count, x, y + 1);
        }
        if (y - 1 >= 0 && !visited[x][y - 1] && isNextValid(board, x, y - 1, word, count))
        {
            findPath(board, word, flag, visited, count, x, y - 1);
        }

        visited[x][y] = 0;
    }

    bool exist(vector<vector<char>> &board, string word)
    {
        bool flag = false;

        if (!checkPresent(board, word))
            return false;

        for (int i = 0; i < board.size(); ++i)
        {
            for (int j = 0; j < board[i].size(); ++j)
            {
                if (board[i][j] == word[0])
                {
                    vector<vector<int>> visited(board.size(), vector<int>(board[0].size(), 0));
                    findPath(board, word, flag, visited, 0, i, j);

                    if (flag == true)
                    {
                        return true;
                    }
                }
            }
        }
        return false;
    }
};